source/common/upstream/load_balancer_impl.h

```diff
@@ -413,15 +416,29 @@ class RoundRobinLoadBalancer : public EdfLoadBalancerBase {
     rr_indexes_.insert({source, seed_});
   }
   double hostWeight(const Host& host) override { return host.weight(); }
+  HostConstSharedPtr unweightedHostPeek(const HostVector& hosts_to_use,
+                                        const HostsSource& source) const override {
+    uint64_t* peekahead_index = const_cast<uint64_t*>(&peekahead_index_);
+    auto i = rr_indexes_.find(source);
+    if (i == rr_indexes_.end()) {
+      return nullptr;
+    }
+    return hosts_to_use[(i->second + (*peekahead_index)++) % hosts_to_use.size()];
+  }
+
   HostConstSharedPtr unweightedHostPick(const HostVector& hosts_to_use,
                                         const HostsSource& source) override {
+    if (peekahead_index_ > 0) {
+      --peekahead_index_;
+    }
     // To avoid storing the RR index in the base class, we end up using a second map here with
     // host source as the key. This means that each LB decision will require two map lookups in
     // the unweighted case. We might consider trying to optimize this in the future.
     ASSERT(rr_indexes_.find(source) != rr_indexes_.end());
     return hosts_to_use[rr_indexes_[source]++ % hosts_to_use.size()];
   }
 
+  uint64_t peekahead_index_{};
```

Comment:
IIRC EDS updates trigger result in a clear + recomputation

Reply:
I think we are keeping a single peekahead_index_ per LB instead of separate peekahead_index_ per scheduler.  For example, prefetch may not work as expected if a series of picks ends up switching between schedulers due to callers alternating between a AllHosts pick and a HealthyHosts pick.